Support @Min and @Max constraint for BigDecimal

amondnet opened the pull request:
- `BigDecimal` 매개변수에 `@Min` 애노테이션 적용
- `BigDecimal` 매개변수에 `@Max` 애노테이션 적용

resolve https://github.com/AutoParams/AutoParams/issues/288

amondnet committed:
Support @Min and @Max constraint for BigDecimal

amondnet committed:
Refactor BigDecimalGenerator to improve annotation handling and use ObjectGeneratorBase

gyuwon commented on autoparams/src/main/java/autoparams/generator/BigRandom.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public static BigInteger between(BigInteger min, BigInteger max, Random r) {
+        return min.add(nextBigInteger(max.subtract(min), r));
+    }
+
+    public static BigDecimal between(BigDecimal min, BigDecimal max, Random r) {
+        return min.add(
+            nextBigDecimal(max.subtract(min), Math.max(min.precision(), max.precision()), r));
+    }
+
+    public static BigInteger between(BigInteger min, BigInteger max) {
+        return between(min, max, ThreadLocalRandom.current());
+    }
+
+    public static BigDecimal between(BigDecimal min, BigDecimal max) {
+        return between(min, max, ThreadLocalRandom.current());
+    }
+
+    // Static Methods for generating Random BigDecimal values:
+
+    public static BigDecimal nextBigDecimal(int scale) {
+        return nextBigDecimal(scale, ThreadLocalRandom.current());
+    }
+
+    public static BigDecimal nextBigDecimal(int scale, Random r) {
+        BigInteger bi = nextBigInteger(scale,
+            r);  // generate random BigInteger with a number of digits equal to scale.
+        BigDecimal bd = new BigDecimal(bi);  // convert BigInteger to a BigDecimal
+        return bd.movePointLeft(bd.precision());  // move the decimal point all the way to the left
+    }
+
+    public static BigDecimal nextBigDecimal(BigDecimal norm, int scale) {
+        return nextBigDecimal(norm, scale, ThreadLocalRandom.current());
+    }
+
+    public static BigDecimal nextBigDecimal(BigDecimal norm, int scale, Random r) {
+        return norm.multiply(nextBigDecimal(scale, r),
+            new MathContext((norm.precision() - norm.scale()) + scale));
+    }
```
이 메서드들 중 실제로 사용되는 코드는 어떤 것인가요?

gyuwon replied:
@amondnet 고맙습니다~

gyuwon commented on autoparams/src/main/java/autoparams/generator/BigDecimalGenerator.java:
```diff
@@ -1,28 +1,47 @@
+        return BigRandom.between(BigDecimal.valueOf(min), BigDecimal.valueOf(max));
     }
 
     private static int getBound() {
         return 1000000 + 1;
     }
```
이번 변경 내용에 해당되지 않음.

amondnet committed:
Make BigRandom class package-private

Reduce visibility from public to package-private as this class is
intended for internal use within the generator package.

amondnet committed:
Fix exception type in BigDecimalGenerator

Replace IllegalArgumentException with RuntimeException to match test
expectations when @Max constraint is less than @Min constraint.

amondnet committed:
Remove unused BigInteger methods from BigRandom

Remove BigInteger-related methods that are not used by the current
BigDecimalGenerator implementation to keep the code focused on its
current purpose.

gyuwon commented on autoparams/src/main/java/autoparams/generator/BigRandom.java:
```diff
@@ -0,0 +1,44 @@
+        return (int) (n / log2);
+    }
+
+    public static BigDecimal between(BigDecimal min, BigDecimal max, Random r) {
+        return min.add(
+            nextBigDecimal(max.subtract(min), Math.max(min.precision(), max.precision()), r));
+    }
+
+    public static BigDecimal between(BigDecimal min, BigDecimal max) {
+        return between(min, max, ThreadLocalRandom.current());
+    }
+
+    // Static Methods for generating Random BigDecimal values:
+
+    public static BigDecimal nextBigDecimal(int scale) {
+        return nextBigDecimal(scale, ThreadLocalRandom.current());
+    }
+
+    public static BigDecimal nextBigDecimal(int scale, Random r) {
+        BigInteger bi = new BigInteger(bitCount(scale), r);
+        BigDecimal bd = new BigDecimal(bi);  // convert BigInteger to a BigDecimal
+        return bd.movePointLeft(bd.precision());  // move the decimal point all the way to the left
+    }
+
+    public static BigDecimal nextBigDecimal(BigDecimal norm, int scale, Random r) {
+        return norm.multiply(nextBigDecimal(scale, r));
+    }
```
이 메서드들 중 public으로 노출되어야 하는 메서드는 어떤 것인가요?

amondnet replied:
수정했습니다. BigDecimalGenerator에서 실제로 사용하는 between(BigDecimal, BigDecimal) 메서드만 package-private으로 노출하고, 나머지 헬퍼 메서드들은 모두 private으로 변경했습니다.

amondnet committed:
Refactor BigRandom to improve encapsulation

Adjusted method visibility to expose only the necessary API and fixed
the logarithm calculation for improved accuracy.

amondnet commented on autoparams/src/main/java/autoparams/generator/BigRandom.java:
```diff
@@ -0,0 +1,44 @@
+class BigRandom {
+
+    // Constants:
+    private static final double log2 = Math.log10(2.0);
```
수정했습니다. log2 상수를 제거하고 bitCount 메서드에서 직접 올바른 계산식을 사용하도록 변경했습니다.

amondnet commented on autoparams/src/main/java/autoparams/generator/BigRandom.java:
```diff
@@ -0,0 +1,44 @@
+    // Computes number of bits needed to represent an n digit positive integer.
+
+    private static int bitCount(int n) {
+        return (int) (n / log2);
```
수정했습니다. 제안해주신 대로 Math.ceil(n * Math.log(10) / Math.log(2))를 사용하도록 변경했습니다.
